genvalidity/src/Data/GenValidity.hs

```diff
@@ -118,6 +148,20 @@ import Control.Monad (forM)
 --
 -- > instance GenUnchecked MyType where
 -- >     genUnchecked = MyType <$> genUnchecked <*> genUnchecked
+--
+--
+-- __Step 2__: If an instatiation via 'Generic' is not possible, then you should emulate what
+--         'genericGenUnchecked' does.
+--         This means that all sub-parts should be  generated using 'genUnchecked'.
+--         Make sure to generate any possible value, valid or not, that can exist at runtime
+--         even when taking the existance of 'Unsafe.Coerce.unsafeCoerce' into account.
+--
+-- === Warning: Invalid values can be funky
+--
+-- Some types have serious validity constraints. See 'Text' or 'ByteString' for example.
+-- These can behave very strangely when they are not valid.
+-- In that case, __do not override' GenUnchecked' to generate only valid values__.
+-- Instead, use 'GenValid' (see below) instead.
```

Comment:
This doesn't make sense:
```
-- In that case, __do not override' GenUnchecked' to generate only valid values__.
-- Instead, use 'GenValid' (see below) instead.
```
`GenUnchecked` is a super class of `GenValid`, so you have to implement `GenUnchecked` somehow before you can use `GenValid`. Keep in mind that `Generic` instance is not always available, so you gotta write it manually.

Reply:
I guess that what I wrote is a bit ambiguous. Is this clearer:

```
-- In that case, __do not override' GenUnchecked' such that 'genUnchecked' only generates valid values__.
-- Instead, use 'GenValid' if you want valid to generate values (see below).
```

Reply:
Not really :) I think it got even more confusing :)
Maybe instead of "do not override' GenUnchecked' such that ..." you should say what should be done with that instance, cause it has to exist one way or another, right?
Something like: "In that case, __override' GenUnchecked', such that it generates ..."

Reply:
"In that case, do not override `genUnchecked`, just use `GenValid` instead."?

Reply:
Also the tick is off for `'GenUnchecked'` vs `' GenUnchecked'`. Note the whitespace.

Reply:
But you have to implement it if there is no `Generic` instance for the datatype, right?

Reply:
> But you have to implement it if there is no Generic instance for the datatype, right?

Yes, I can try to explain that further.
The semantics-preserving way to do that is to write a  `genUnchecked` implementation just like `Generic` would:
```
FooBar <$> genUnchecked <*> genUnchecked ...
```
... and then _not_ use `genUnchecked`.

Reply:
But I'm seriously considering bending the principles in the case of `Text` and `ByteString`. The thing is that all sorts of weird stuff happens when invariants are not preserved. Even with `Rational`, and that's exactly what `Validity` *is for*.

Reply:
That explanation certainly clarifies things, but yields `GenUnchecked` useless for anything that uses  `ByteString` or `Text`.

Reply:
> That explanation certainly clarifies things

Great, but I had hoped that my docs could do that :P Could you help me fix the docs so that they would clarify what they didn't?

> but yields GenUnchecked useless for anything that uses ByteString or Text.

Yes that's right. Keep in mind that the library optimises for removing friction in the typical case.
You can still implement `genUnchecked = undefined` and `genValid = genValidStructurally` in that case and not worry about this.

Reply:
For example in case of above `data FooBar = FooBar ByteString ...`, it might be that it can contain values that are invalid, as far as `FooBar` is concerned, but valid from the point of `ByteString`. In which case you would want to do `genUnchecked = FooBar <$> genValid <*> genUnchecked ...`, but that goes against what you are suggesting.

Reply:
> For example in case of above data FooBar = FooBar ByteString ..., it might be that it can contain values that are invalid, as far as FooBar is concerned, but valid from the point of ByteString. 

That does not make sense because then the first `ByteString` *is already checked to be valid* so the value is no longer entirely unchecked.

Reply:
(The opposite _is_ possible by the way. It could be that `newtype Wut = Wut Rational` is valid even if `Rational` is not if you define `Validity` in such a way.)

Reply:
I think what you're getting at is that there could be multiple levels at which a valid can be invalid. Sadly there isn't much support for distinguishing between those.

Reply:
> The thing is that all sorts of weird stuff happens when invariants are not preserved. Even with Rational, and that's exactly what Validity is for.

I suspect that with `Rational` it doesn't result in segfaults ;) Also if you say don't use this instance, you preventing any weird stuff from being discovered.

Reply:
> I suspect that with Rational it doesn't result in segfaults ;) 

That's right, but in some sense at least points you to the fact that there **is** a problem, whereas this doesn't:
```
> 5 %: 5  ==  1 %: 1
False
```

Reply:
> Also if you say don't use this instance, you preventing any weird stuff from being discovered.

That depends on whether your code needs to deal with invalid values.
If your values can be invalid in multiple ways, best to use specific generators instead of `genInvalid` (or `genUnchecked`) .

Reply:
> but in some sense a segfault at least points you to the fact that there is a problem

But it does not point you to where this problem is even remotely. You don't even know which test case caused the segafault.

>  If your values can be invalid in multiple ways, best to use specific generators instead of genInvalid (or genUnchecked) .

I guess this is what it boils down to. You cannot use `GenUnchecked` or `GenInvalid` you are are left with custom generators of invalid values for `ByteString`, `Text` and all `Vector`s.

Reply:
> I guess this is what it boils down to. You cannot use GenUnchecked or GenInvalid you are are left with custom generators of invalid values for ByteString, Text and all Vectors.

Yes, that's right I think.

Reply:
> Yes, that's right I think.

IMHO it's a bad design decision, but again, that is your library, you have your own vision about.

I'll tell you one thing for certain though, I will refuse to have any library as a dependency, even it is  `genvalidity-*` and is used just for testing, if it can produce segfaults.

Reply:
> IMHO it's a bad design decision, but again, that is your library, you have your own vision about.

So let's talk about alternatives. `error "don't use this"` is an option as far as I'm concerned. Would that  be better than what's there?

Reply:
Absolutely! A good and concise explanation if why it should not be used would very helpful as well, but that is besides the point.

Reply:
At least with an error you will see which test case is failing and why, instead of resorting to debugging every unsafe operation in the codebase.

Reply:
> A good and concise explanation if why it should not be used would very helpful as well, but that is besides the point.

Right, it would say "use genValid instead" and point to a document that explains the situation